pallets/pool-system/src/tranches.rs

```diff
@@ -1133,14 +1190,26 @@ where
 		I: IntoIterator<Item = W>,
 	{
 		let mut res = Vec::with_capacity(self.tranches.len());
-		let iter = self.tranches.iter().rev().zip(with.into_iter());
-
-		for (tranche, w) in iter {
-			let r = f(tranche, w)?;
-			res.push(r);
+		let mut tranche_slice = self.non_residual_top_slice().iter();
+		let mut with_iter = with.into_iter();
+
+		for _ in 0..self.tranches.len() {
```

Comment:
Question: Do we need to distinguish between errors based on different lengths of `tranche_slice ` and `with_iter`? If not, we can revert back to zipping which automatically halts when the first iterator stopps. cc @offerijns 

If we care, I would like this custom `zipping` to be extracted to a function to reduce duplicate code as it is called thrice.

Reply:
> If we care, I would like this custom zipping to be extracted to a function to reduce duplicate code as it is called thrice.

Yes we care. Would be nice to extract.

Reply:
Comparing length would also be fine.

Reply:
Seems like @thea-leake's approach is optimal here actually!

> Comparing length would also be fine.
Even though comparing input lengths is quicker and more readable, it requires adding the `Clone` trait to the iterator `I` because counting the elements in the iterator consumes it: 

```rustlang
let with_count = with.clone().into_iter().count();

match self.tranches.len().cmp(&with_count) {) {
    Ordering::Less => Err(DispatchError::Other(
        "Iterable contains more elements than EpochExecutionTranches tranche count",
    )),
    Ordering::Greater => Err(DispatchError::Other(
        "EpochExecutionTranches contains more tranches than iterables elements",
    )),
    Ordering::Equal => {
        self.non_residual_top_slice().iter().zip(with.into_iter()).try_fold(Vec::with_capacity(self.tranches.len()), |mut res, (tranche, w)| {
            res.push(f(tranche, w)?);
            Ok(res)
        })
    }
}
```

Extracting the _duplicate_ code also is not really possible due to the different assumptions of the tranche slize (either `IterMut` or `Rev<IterMut<...>>`) and `f` (either `mut` or not).